presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -929,4 +1015,125 @@ public boolean equals(Object obj)
                     Objects.equals(this.sortChannel, other.sortChannel);
         }
     }
+
+    private static class IsDistinctFromGenerator
+    {
+        private final FunctionRegistry registry;
+
+        public IsDistinctFromGenerator(FunctionRegistry registry, TypeManager typeManager)
+        {
+            this.registry = requireNonNull(registry, "registry is null");
+            for (Type type : typeManager.getTypes()) {
+                try {
+                    operators.getUnchecked(type);
+                }
+                catch (Exception e) {
+                    //ignore
+                }
+            }
+        }
+
+        private final LoadingCache<Type, CacheValue> operators = CacheBuilder.newBuilder()
+                .recordStats()
+                .maximumSize(1000)
+                .build(CacheLoader.from(this::fromKey));
+
+        private CacheValue fromKey(Type key)
+        {
+            Signature signature = registry
+                    .resolveOperator(OperatorType.IS_DISTINCT_FROM, ImmutableList.of(key, key));
+            ScalarFunctionImplementation function = registry.getScalarFunctionImplementation(signature);
+            return new CacheValue(signature.getName(), function.getMethodHandle());
+        }
+
+        private static class CacheValue
+        {
+            private final String name;
+            private final MethodHandle methodHandle;
+
+            public CacheValue(String name, MethodHandle methodHandle)
+            {
+                this.name = requireNonNull(name, "name is null");
+                this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
+            }
+
+            public String getName()
+            {
+                return name;
+            }
+
+            public MethodHandle getMethodHandle()
+            {
+                return methodHandle;
+            }
+        }
+
+        private BytecodeBlock generateIsDistinctFrom(CallSiteBinder callSiteBinder,
+                                                            Type type,
+                                                            BytecodeExpression leftBlock,
+                                                            BytecodeExpression leftBlockPosition,
+                                                            BytecodeExpression rightBlock,
+                                                            BytecodeExpression rightBlockPosition)
+        {
+            BytecodeExpression typeExpression = constantType(callSiteBinder, type);
+            CacheValue operator = operators.getUnchecked(type);
+            Binding binding = callSiteBinder.bind(operator.getMethodHandle());
+
+            BytecodeBlock block = new BytecodeBlock().setDescription("invoke");
+
+            if (type.getJavaType() == Block.class && !(type instanceof ArrayType)) {
```

Comment:
Use `BytecodeUtil.generateInvocation` instead of re-inventing Presto function calling convention here. That will allow you to avoid special casing map type and row type here.